cdap-app-fabric/src/test/java/co/cask/cdap/gateway/handlers/OperationsDashboardHttpHandlerTest.java

```diff
@@ -16,105 +16,257 @@
+
+  private static ProgramHeartbeatDataset programHeartbeatDataset;
+  private static TransactionExecutor heartBeatTxnl;
 
   @BeforeClass
-  public static void setup() {
-    store = getInjector().getInstance(DefaultStore.class);
+  public static void setup() throws Exception {
+    Injector injector = getInjector();
+    TransactionExecutorFactory txExecutorFactory =
+      injector.getInstance(TransactionExecutorFactory.class);
+    DatasetFramework datasetFramework = injector.getInstance(DatasetFramework.class);
+    DatasetId heartbeatDataset = NamespaceId.SYSTEM.dataset(Constants.ProgramHeartbeat.TABLE);
+    Table heartbeatTable = DatasetsUtil.getOrCreateDataset(datasetFramework, heartbeatDataset, Table.class.getName(),
+                                                           DatasetProperties.EMPTY, Collections.emptyMap());
+    programHeartbeatDataset = new ProgramHeartbeatDataset(heartbeatTable);
+    heartBeatTxnl = txExecutorFactory.createExecutor(Collections.singleton(programHeartbeatDataset));
+  }
+
+  /**
+   * writes heart beat messages starting from startTime + interval up to endTime, each heartbeat separated by interval
+   */
+  private void setUpProgramHeartBeats(RunRecordMeta runRecordMeta,
+                                      long startTime, long endTime, long interval)
+    throws TransactionFailureException, InterruptedException {
+    for (long time = startTime + interval; time < endTime; time += interval) {
+      writeRunRecordMeta(runRecordMeta, time);
+    }
+  }
+
+  private void writeRunRecordMeta(RunRecordMeta runRecordMeta,
+                                  long timestampInMillis) throws InterruptedException, TransactionFailureException {
+    heartBeatTxnl.execute(() -> {
+      programHeartbeatDataset.writeRunRecordMeta(runRecordMeta, timestampInMillis);
+    });
+  }
+
+  /**
+   * setup and return mock program properties with status running, runId and time when it started running are
+   * specified by parameters
+   */
+  private RunRecordMeta getMockRunRecordMeta(NamespaceId namespaceId, RunId runId, StatusTime statusTime,
+                                             @Nullable TriggeringScheduleInfo triggeringScheduleInfo) {
+    ProgramId programId = namespaceId.app("someapp").program(ProgramType.SERVICE, "s");
+    RunRecordMeta.Builder runRecordBuilder = RunRecordMeta.builder();
+    runRecordBuilder.setArtifactId(ARTIFACT_ID.toApiArtifactId());
+    runRecordBuilder.setPrincipal("userA");
+    runRecordBuilder.setProgramRunId(programId.run(runId));
+
+    Map<String, String> systemArgs = new HashMap<>();
+    if (triggeringScheduleInfo != null) {
+      systemArgs.put(ProgramOptionConstants.TRIGGERING_SCHEDULE_INFO, GSON.toJson(triggeringScheduleInfo));
+    }
+    runRecordBuilder.setSystemArgs(systemArgs);
+    runRecordBuilder.setStartTime(RunIds.getTime(runId, TimeUnit.SECONDS));
+    if (statusTime.getProgramStatus().isEndState()) {
```

Comment:
instead of passing in a StatusTime object and having these comparisons, you can make this return a RunRecordMeta.Builder and have the caller set the relevant time. It will make it easier to read and understand by removing the need to jump across methods to understand what is happening, and removes the need for the StatusTime class.

Reply:
updated it, thanks 👍